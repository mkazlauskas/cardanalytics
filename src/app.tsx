
import { levels, setLevel } from 'loglevel';
import React from 'react';
import ReactDOM from 'react-dom';
import { Provider } from 'react-redux';
import Components from './components';
import store from './redux/store';

setLevel(process.env.NODE_ENV === 'production' ? levels.INFO : levels.DEBUG);

const app = <Provider store={store}><Components /></Provider>;

ReactDOM.render(app, document.getElementById('root'));

// Simulating load time. TODO: Move this to an epic
setTimeout(() => {
	document.getElementById('preloader')?.remove();
}, 1000);