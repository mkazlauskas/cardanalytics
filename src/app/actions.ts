import { ActionType, createAction } from 'typesafe-actions';
import { GeneralStats } from './types';

export const loadGeneralStats = createAction('LOAD_GENERAL_STATS')<GeneralStats>();

export const actions = { loadGeneralStats };
export type Action = ActionType<typeof actions>;