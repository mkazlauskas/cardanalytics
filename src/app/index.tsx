
import { createTheme, CssBaseline, MuiThemeProvider } from '@material-ui/core';
import { grey } from '@material-ui/core/colors';
import { levels, setLevel } from 'loglevel';
import React from 'react';
import ReactDOM from 'react-dom';
import { Provider } from 'react-redux';
import { Store } from 'redux';
import Components from '../components';

setLevel(process.env.NODE_ENV === 'production' ? levels.INFO : levels.DEBUG);

const theme = createTheme({ palette:{ background: { default: grey[200] } } });

export default (store: Store) => {
  const app = (
    <Provider store={store}>
      <MuiThemeProvider theme={theme}>
        <CssBaseline />
        <Components />
      </MuiThemeProvider>
    </Provider>
  );

  ReactDOM.render(app, document.getElementById('root'));

  // Simulating load time. TODO: Move this to an epic
  setTimeout(() => {
		document.getElementById('preloader')?.remove();
  }, 1000);
};