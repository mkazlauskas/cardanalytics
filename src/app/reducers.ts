import { combineReducers } from 'redux';
import { getType, StateType } from 'typesafe-actions';
import { Action, loadGeneralStats } from './actions';
import { GeneralStats } from './types';

const generalStats = (state: GeneralStats | null = null, action: Action) => {
  if (action.type === getType(loadGeneralStats)) {
    return action.payload;
  }
  return state;
};

export const rootReducer = combineReducers({ generalStats });

export type State = StateType<typeof rootReducer>;