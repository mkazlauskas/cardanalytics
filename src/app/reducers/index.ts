import { combineReducers } from 'redux';
import { getType, StateType } from 'typesafe-actions';
import { Action, loadCardanoData } from '../actions';
import { CardanoStats } from '../types';

export const cardanoData = (state: CardanoStats | null = null, action: Action) => {
  if (action.type === getType(loadCardanoData)) {
    return action.payload;
  }
  return state;
};

export const rootReducer = combineReducers({ cardanoData });

export type State = StateType<typeof rootReducer>;