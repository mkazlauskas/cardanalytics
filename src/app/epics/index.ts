import { warn } from 'loglevel';
import { Epic } from 'redux-observable';
import { catchError, combineLatest, EMPTY, exhaustMap, filter, map, mergeMap, take, tap, timer } from 'rxjs';
import { isActionOf } from 'typesafe-actions';
import { Action, loadCardanoData } from '../actions';
import { State } from '../reducers';
import { Services } from '../services';

type CardanalyticsEpic = Epic<Action, Action, State, Services>;
export const FETCH_INTERVAL = 10000;

export const updateData$: CardanalyticsEpic = (_$, __$, services) => 
  timer(0, FETCH_INTERVAL).pipe(exhaustMap(() =>
    services.load$().pipe(
      map((data) => loadCardanoData(data)),
      catchError((e) => {
        warn(`Failed to fetch data, retrying in ${FETCH_INTERVAL}ms`, e);
        return EMPTY;
      }),
    )
  ));

export const removePreloader$: CardanalyticsEpic = (action$) => combineLatest([
  action$.pipe(filter(isActionOf(loadCardanoData)))
]).pipe(
  take(1),
  tap(() => document.getElementById('preloader')?.remove()),
  mergeMap(() => EMPTY)
);

export default [updateData$, removePreloader$];