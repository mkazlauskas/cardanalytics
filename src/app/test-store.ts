import { createStore } from 'redux';
import { State } from './reducers';
import { CardanoStats } from './types';

export const testCardanoData: CardanoStats = {
  blocks: 5995095,
  transactions: 9787836,
  assets: 43751,
  epoch: 278,
  slot: 300948,
}; 

const store = createStore(
  (): State => ({
    cardanoData: testCardanoData
  }),
);

export default store;

// Simulating load time.
setTimeout(() => {
	document.getElementById('preloader')?.remove();
}, 1000);