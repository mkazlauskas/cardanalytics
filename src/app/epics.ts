import { Epic } from 'redux-observable';
import { of } from 'rxjs';
import { Action, loadGeneralStats } from './actions';
import { State } from './reducers';
import { Services } from './services';
import { generalStats } from './test-store';

type CardanalyticsEpic = Epic<Action, Action, State, Services>;

// TODO: replace with loading actual stats, not test stats
const load$: CardanalyticsEpic = () => of(loadGeneralStats(generalStats));

export default [load$];