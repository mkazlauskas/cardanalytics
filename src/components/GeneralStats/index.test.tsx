import React from 'react';
import { useSelector } from 'react-redux';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Component from '.';
import { generalStats } from '../../app/test-store';

describe('GeneralStats', () => {
  beforeAll(() => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (useSelector as any).mockImplementation(() => generalStats);
  });
  it('should match snapshot', () => {
    expect(toJson(shallow(<Component />))).toMatchSnapshot();
  });
});