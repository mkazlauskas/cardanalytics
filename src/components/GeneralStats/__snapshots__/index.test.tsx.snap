// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GeneralStats should match snapshot 1`] = `
<WithStyles(ForwardRef(Grid))
  container={true}
  direction="column"
>
  <span>
    Total blocks:
     
    5,995,095
  </span>
  <span>
    Total transactions:
     
    9,787,836
  </span>
  <span>
    Total stake pools:
     
    2,877
  </span>
  <span>
    Total ADA stake:
     
    23,240,000,000
  </span>
  <span>
    Total assets:
     
    43,751
  </span>
  <span>
    Era:
     
    Shelley
  </span>
  <span>
    Epoch:
     
    278
  </span>
  <span>
    Slot:
     
    300,948
  </span>
</WithStyles(ForwardRef(Grid))>
`;
